### PR TITLE
feat: provide an illustration component

### LINKED ISSUE #157: Component: Icon Illustrations
### Design
https://www.figma.com/file/FWlI63N28LP7BxqNqj1bGw/sy-004-(Design!)?node-id=3659%3A25045

### Things to consider
- Which of the existing Icons is still of any use?
- In Figma there seem to be mostly unique names but there are already duplicates (like "satellite"). How do we deal with this?

### Expected outcome
A new component that can be used to show any of the given illustration icons.

### PR DESCRIPTION
**Pull request by georgiee**
close #157 

---
## Description
I have manually saved all illustrations from Figma, processed the file content and names slightly to expose them easily with SVGR through a single file. Being there I decided to create an `<Illustration/>` component with a typed property to pick the one you want to display and not to fiddle around with components.

Once this worked I created a special story to showcase all available illustrations in an overview with a search field.

And because this proofs nothing I tried to use the Illustration component in the Callout component. I refactored it to accept a string instead of a component. The string has the same typing as the illustrations. This makes the Callout basically dependent on the Illustration component, which is what it is I think.

You can now use illustrations in Callouts the same way as with the Illustration component itself:

```
<Callout illustration="earth_010"/>
<Illustration show="earth_010"/>
```

I'm not sure about the `show` prop name, sounds more like a boolean.

## Preview
Screen Shot 2021-11-16 at 14 34 51-fullpage

Screen Shot 2021-11-16 at 14 34 35-fullpage

## API
1. There is a closed set of available illustrations. The prop `show` is typed accordingly. 
Screen Shot 2021-11-16 at 14 40 36
2. You can set three different sizes: 72px (normal, default), 144px (large) and inherit which will let the illustration fit its container (basically width 100%).

## Remarks
1. I tried to adapt the Leadbox, but the current css/html architecture is  not suited for a quick change, so let's update it in a separate PR.
2. The jsdoc extraction works for this component and displays my comment in the docs:
Screen Shot 2021-11-16 at 14 43 33

Also the props are partially extracted. It looks better outside of the docs but I think there are some nasty problems hidden since this extraction is a new feature.

### CODE REVIEW

File: src/components/illustration/illustration-files.tsx
Comment (georgiee) [inline src/components/illustration/illustration-files.tsx]: I deliberately decided to split this file from  the export `illustration-set.tsx` where I do the types & keys processing so we can easily update this file without worrying about actual code that does things.

File: src/components/illustration/illustration.stories.tsx
Comment (georgiee) [inline src/components/illustration/illustration.stories.tsx]: I had argTypes defined here, but they really don't coexist well with the automatically extracted values. Plus they seem to bleed into any story, even if excluded. So weird. I removed it here  as the automatic extraction works okay for the component itself and I define only a single field for the overview story later.

File: src/components/illustration/illustration.stories.tsx
Comment (georgiee) [inline src/components/illustration/illustration.stories.tsx:29]: I have no clue if this is best practice. It felt like I'm doing something forbidden, but it's so useful. Why wouldn't I add this? Maybe there is a story type that is more isolated? I couldn't find any matching pattern in the docs though.

https://user-images.githubusercontent.com/1701755/141997289-c75d1218-ecd4-46ed-9bfe-6a8215666676.mp4

File: src/components/illustration/illustration.stories.tsx
Comment (georgiee) [inline src/components/illustration/illustration.stories.tsx]: This the the actual component related story we already know. And by the way the value is set as the correct default in the generated list. Did not work before with manual argTypes.

Screen Shot 2021-11-16 at 14 51 38

File: src/components/illustration/illustration.tsx
Comment (georgiee) [inline src/components/illustration/illustration.tsx]: I wrote a few jsdocs comments here. And I swear, they were visible in the automatically generated table in `Docs` at some point (hence the motivation to write them).

Now it looks like this:
Screen Shot 2021-11-16 at 14 52 23

`Search` is wrong and from the overview component. While writing this comment I tried to swap the order and the table updated. So it's a hidden convention that the first story is taken for the table generation I guess. 
Screen Shot 2021-11-16 at 14 53 42

I push the new order. This will remove the overview as the first story which was kind of a good UX I think. Anyway.

File: src/components/callout/callout.stories.tsx
Comment (georgiee) [inline src/components/callout/callout.stories.tsx]: I removed the manual mapping and I simpky provide the available illustrations which is then passed to the according parameter. 
Screen Shot 2021-11-16 at 14 57 15

File: src/components/callout/callout.tsx
Comment (georgiee) [inline src/components/callout/callout.tsx:39]: i remove the explicit sizing to rely on the illustration being sized appropriately. I removed the smaller variant,
as we don't have such a size for illustrations. It still looked good with that size. I would like to take that easy route as the 48px was a blind guess anyway before.


Screen Shot 2021-11-16 at 15 00 12

File: src/components/callout/callout.stories.tsx
Comment (fischeversenker) [inline src/components/callout/callout.stories.tsx]: Now there is no way to select `None`. Could you re-add that?

File: src/components/illustration/illustration-files.tsx
Comment (fischeversenker) [inline src/components/illustration/illustration-files.tsx]: It seems very important to you to stress how much manual effort this was. Thanks a lot for taking care :pray:
However, I'm not sure how much value this comment provides... do you insist in keeping it?

File: src/components/illustration/illustration-files.tsx
Comment (fischeversenker) [inline src/components/illustration/illustration-files.tsx:51]: I personally would prefer the schema `123_name`, e.g. would prefer `051_venus` over `venus_051`. What made you turn around the number and name?

File: src/components/illustration/illustration.stories.tsx
Comment (fischeversenker) [reply src/components/illustration/illustration.stories.tsx]: If you want these `argTypes` only on the `Regular` story you could add them there instead of here in this "global" config.

File: src/components/illustration/illustration.stories.tsx
Comment (fischeversenker) [inline src/components/illustration/illustration.stories.tsx]: I'm assuming this last line is outdated?

File: src/components/illustration/illustration.stories.tsx
Comment (fischeversenker) [inline src/components/illustration/illustration.stories.tsx]: Since we don't want this control to appear on the Docs page I'm wondering if we can remove this `table` property?

File: src/components/illustration/illustration.stories.tsx
Comment (fischeversenker) [inline src/components/illustration/illustration.stories.tsx]: This sentence misses the

File: src/components/illustration/illustration.tsx
Comment (fischeversenker) [inline src/components/illustration/illustration.tsx]: Large, according to the comments in `IllustrationSize` and the css, is `144x144px`. Would you please update this?

File: src/components/illustration/illustration.stories.tsx
Comment (fischeversenker) [reply src/components/illustration/illustration.stories.tsx:29]: I like it! I agree that it feels slightly hacky but it works and provides great value so let's keep it 🙌

File: src/components/illustration/illustration.tsx
Comment (fischeversenker) [reply src/components/illustration/illustration.tsx]: I feel like "Regular" should be the first story for every component so I'm fine with the now-changed order. That undocumented convention regarding the order affecting the table generation is nice to know.

File: src/components/illustration/illustration.stories.tsx
Comment (fischeversenker) [inline src/components/illustration/illustration.stories.tsx]: There is only one such story and the next template is introduced without a proper separator-comment so this comment is slightly misleading.

File: src/components/illustration/illustration-set.ts
Comment (fischeversenker) [inline src/components/illustration/illustration-set.ts:10]: Very useful type magic in here. Approved 👍

File: src/components/callout/callout.tsx
Comment (fischeversenker) [inline src/components/callout/callout.tsx]: Why did you decide to export this?

File: src/components/illustration/illustration.tsx
Comment (fischeversenker) [inline src/components/illustration/illustration.tsx:22]: This usage of `Pick<>` is impressive... but seems like overkill here to be honest. What's the benefit of keeping this?

**Review by fischeversenker — COMMENTED** [review]
Thanks for taking care of this. Especially the laborious extraction of the svgs from Figma :pray:

I left a bunch of comments and I'm also not particularly fond of the property name `show`... I think I understand the API-driven idea that led to `show` but it sounds wrong. I can't think of anything nicer right now (except `illustration` which obviously feels very redundant) so I'm fine with keeping `show` for now.

File: src/components/illustration/illustration.stories.tsx
Comment (georgiee) [reply src/components/illustration/illustration.stories.tsx:29]: Thanks 🙏

File: src/components/illustration/illustration.tsx
Comment (georgiee) [reply src/components/illustration/illustration.tsx]: makes sense and people will come along I guess 😎

File: src/components/illustration/illustration-files.tsx
Comment (georgiee) [reply src/components/illustration/illustration-files.tsx]: No, we can remove it. I think it's important to note two things though:
+ When was this file created (I know git blame can tell, but why not being explicit here for readability?)
+ That this is a manual file and not generated, which people might assume?

Basically I fear updates in the future and that we don't know how we created this file in the first place. I'm happy to remove, but on the other side, nobody sees this file anyway in the best case so the additional information doesn't hurt, does it ?

File: src/components/callout/callout.stories.tsx
Comment (georgiee) [reply src/components/callout/callout.stories.tsx]: good point I will provide this again then.

File: src/components/illustration/illustration-files.tsx
Comment (georgiee) [reply src/components/illustration/illustration-files.tsx:51]: Because you can't start variable names with numbers 🤓

File: src/components/callout/callout.stories.tsx
Comment (georgiee) [reply src/components/callout/callout.stories.tsx]: fixed and I took the chance to reduce the variants as it doesn't help in this case anyway.

Screen Shot 2021-11-17 at 06 02 16

File: src/components/illustration/illustration-files.tsx
Comment (georgiee) [reply src/components/illustration/illustration-files.tsx]: Removed.

File: src/components/illustration/illustration.tsx
Comment (georgiee) [reply src/components/illustration/illustration.tsx:22]: I only use the property `show` in this definition. Would you expect me to use an isolated type with a single property name or just pass in the `IllustrationProps`?

File: src/components/callout/callout.tsx
Comment (georgiee) [reply src/components/callout/callout.tsx]: One desperate attempt to render the docs correctly. I though the extractor needs this to be exported (having some feeling from my experiences with the introspection utility in our customer project). I will revert this of course, sorry this slipped in.

File: src/components/illustration/illustration-set.ts
Comment (georgiee) [reply src/components/illustration/illustration-set.ts:10]: Thanks!

What I'm still missing is a typed hash. Right now it's easy to use the typed props
but when UI want to access it somewhere else, like in the story in only have this big fat array of keys `ILLUSTRATION_NAMES` which I can access `ILLUSTRATION_NAMES[0]` but I would rather have something like the original export `ILLUSTRATION_HASH.blimp_001` (which should then return a string `blimp_001`) and such. I had a reduce in there but I wasn't able to quickly type it appropriately. So there is a challenge left 😁🤘

File: src/components/illustration/illustration.stories.tsx
Comment (georgiee) [reply src/components/illustration/illustration.stories.tsx]: Updated. Thanks.

File: src/components/illustration/illustration.tsx
Comment (georgiee) [reply src/components/illustration/illustration.tsx]: Which is the correct size. I think the 72 bleed into my thoughts, thanks 😬

File: src/components/illustration/illustration.stories.tsx
Comment (georgiee) [reply src/components/illustration/illustration.stories.tsx]: updated to this. thanks.

> We only want to expose our `filter` property on this special story

File: src/components/illustration/illustration.stories.tsx
Comment (georgiee) [reply src/components/illustration/illustration.stories.tsx]: Oh, indeed. Sneaky debugging leftovers. Thank you again for your attentive eye.

File: src/components/illustration/illustration.tsx
Comment (georgiee) [inline src/components/illustration/illustration.tsx]: leftover from tinkering round with my component it the lead component context where I tried to style our Illustration component with styled components which forwards a `className`. We still might want to support styled components by adding a className to our components but that's a separate discussion.

**Comment by georgiee** [comment]
Thank you @fischeversenker for your thorough review.

I have pushed a commit which handles all of the issues:
https://github.com/satellytes/satellytes.com/pull/190/commits/1a5b37dbb25f62da33a95731346a4367ce2864e0

Once checked, I hope this PR is ready to merge 🙌

File: src/components/illustration/illustration.tsx
Comment (georgiee) [reply src/components/illustration/illustration.tsx:22]: By the way, I didn't want to randomly use an impressive typescript tool. `Pick` and the counterpart `Omit` are some great utility types to know and use (wisely) in your daily work 💪

Reference: https://www.typescriptlang.org/docs/handbook/utility-types.html

File: src/components/illustration/illustration.stories.tsx
Comment (georgiee) [reply src/components/illustration/illustration.stories.tsx]: Removed. Thanks.

File: src/components/illustration/illustration.stories.tsx
Comment (georgiee) [reply src/components/illustration/illustration.stories.tsx]: I tried this and yes it works. I was curious then. When will the docs engine read it? And what I found is weird.
That parameter is ignored when it's placed on the overview story. I though it's related to the export order but changing this would not work either. It's only accepted on the "normal" story and on the default export. Maybe the `as any` of the overview story makes things worse but at some point I stopped testing this. I will check in future stories again how this behaves.

The more correct place is the default export I think, because that's describing the overall component we are dealing with, so I keep it.

File: src/components/illustration/illustration-files.tsx
Comment (fischeversenker) [reply src/components/illustration/illustration-files.tsx:51]: Alright. Good reason :joy:

File: src/components/illustration/illustration.tsx
Comment (fischeversenker) [reply src/components/illustration/illustration.tsx:22]: I would have expected you to just pass in `IllustrationProps`. I don't see the benefit of using `Pick` here. The props are getting destructured and you are only pulling out the `size` property anyways.
Another point: This styled component will receive all the properties of `IllustrationProps` so this generic type might even be considered wrong since it hides away other properties of `IllustrationProps` that are very well available in the scope of the styled component.
My main concern is that it will confuse more juniory developers and make them wonder why this is used here and nowhere else.

This is clearly not a blocker and I'm fine with keeping it so I'll leave it to you to decide.

**Review by fischeversenker — APPROVED** [review]
Thanks for the cleanup. From my point of view this is ready to be merged now, yes 🚢

File: src/components/illustration/illustration.tsx
Comment (georgiee) [reply src/components/illustration/illustration.tsx:22]: Regarding  this quote:

> might even be considered wrong since it hides away other properties of IllustrationProps that are very well available in the scope of the styled component.

I think we have a misunderstanding. I try to clarify this but thanks for unblocking this issue 👍

---

Here a quick snippet from the related code (simplified)

```
export const Illustration = (props: IllustrationProps) => {
  return (
    <IllustrationLayout size={props.size}> // <-- our styled component with the props in discussion
      <IllustrationSvg />
    </IllustrationLayout>
  );
};
```

You can see that the component `IllustrationLayout` has the following props definition (writte here for demonstration purposes only)

```
interface IllustrationLayout {
   size: IllustrationSize;
}
```

which is different (a actual subset) from the props coming in from the outside, being the full `IllustrationProps`:

```
export interface IllustrationProps {
  show: IllustrationType;
  size?: IllustrationSize;
}
```

That's why this pick make sense. If I would use `IllustrationProps` on the styled component definition (`const IllustrationLayout = styled.div<IllustrationProps>`) I would suggest that there are other properties like `show` which is wrong.

**Comment by georgiee** [comment]
Thanks for the review. I will merge this now 🌟

File: src/components/illustration/illustration.tsx
Comment (fischeversenker) [reply src/components/illustration/illustration.tsx:22]: Got it. You are right. I was assuming (without double checking, my bad) that you just did `{...props}` on this component.

Comment (georgiee) [reply src/components/illustration/illustration.tsx:22]: I did before for the styled components experiment ☝️ Riddle solved 😁

See my explanation:
https://github.com/satellytes/satellytes.com/pull/190#discussion_r750896607

**Pull request merged**
